controlplane/pkg/nsm/nsm.go

```diff
@@ -0,0 +1,411 @@
+			LocalSource: requestConnection.(*connection.Connection),
+		}
+	}
+
+	// We handling request from local or remote endpoint.
+	//TODO: in case of remote NSE( different cluster case, this method should be changed)
+	if !srv.isLocalEndpoint(endpoint) {
+		dpApiConnection.Destination = &crossconnect.CrossConnect_RemoteDestination{
+			RemoteDestination: nseConnection.(*remote_connection.Connection),
+		}
+	} else {
+		dpApiConnection.Destination = &crossconnect.CrossConnect_LocalDestination{
+			LocalDestination: nseConnection.(*connection.Connection),
+		}
+	}
+	return dpApiConnection
+}
+func (srv *networkServiceManager) validateNSEConnection(nseConnection nsm.NSMConnection) error {
+	err := nseConnection.IsComplete()
+	if err != nil {
+		err = fmt.Errorf("failure Validating NSE Connection: %s", err)
+		return err
+	}
+	return nil
+}
+
+func (srv *networkServiceManager) createConnectionId() string {
+	return srv.model.ConnectionId()
+}
+
+func (srv *networkServiceManager) closeDataplane(clientConnection *model.ClientConnection) error {
+	logrus.Info("Closing cross connection on dataplane...")
+	dataplaneClient, conn, err := srv.serviceRegistry.DataplaneConnection(clientConnection.Dataplane)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+	if conn != nil {
+		defer conn.Close()
+	}
+	if _, err := dataplaneClient.Close(context.Background(), clientConnection.Xcon); err != nil {
+		logrus.Error(err)
+		return err
+	}
+	logrus.Info("Cross connection successfully closed on dataplane")
+	return nil
+}
+
+func (srv *networkServiceManager) getNetworkServiceManagerName() string {
+	return srv.model.GetNsm().GetName()
+}
+
+func (srv *networkServiceManager) updateConnectionParameters(nseConnection nsm.NSMConnection, endpoint *registry.NSERegistration) {
+	if srv.getNetworkServiceManagerName() == endpoint.GetNetworkserviceEndpoint().GetEndpointName() {
+		workspace := nsmd.WorkSpaceRegistry().WorkspaceByEndpoint(endpoint.GetNetworkserviceEndpoint())
+		if workspace != nil { // In case of tests this could be empty
+			nseConnection.(*connection.Connection).GetMechanism().GetParameters()[connection.Workspace] = workspace.Name()
+		}
+	}
+}
+
+func (srv *networkServiceManager) updateExcludePrefixes(requestConnection nsm.NSMConnection) {
+	c := requestConnection.GetContext()
+	if c == nil {
+		c = &connectioncontext.ConnectionContext{}
+	}
+	for _, ep := range srv.excluded_prefixes {
+		c.ExcludedPrefixes = append(c.ExcludedPrefixes, ep)
+	}
+	// Since we do not worry about validation, just
+	requestConnection.SetContext(c)
+}
+
+func (srv *networkServiceManager) getEndpoint(ctx context.Context, requestConnection nsm.NSMConnection, ignore_endpoints map[string]*registry.NSERegistration) (*registry.NSERegistration, error) {
+
+	// Handle case we are remote NSM and asked for particular endpoint to connect to.
+	targetEndpoint := requestConnection.GetNetworkServiceEndpointName()
+	if len(targetEndpoint) > 0 {
```

Comment:
What does an empty `targetEndpoint` mean?

Reply:
targetEndpoint is if NSC -> NSM(select target engpoint) -> NSM(use target endpoint by name)